portfolio/src/main/java/com/google/sps/servlets/DoggoRedditServlet.java

```diff
@@ -0,0 +1,136 @@
+import java.util.ArrayList;
+import java.util.Arrays;
+import java.util.List;
+import java.util.Map;
+import javax.servlet.annotation.WebServlet;
+import javax.servlet.http.HttpServlet;
+import javax.servlet.http.HttpServletRequest;
+import javax.servlet.http.HttpServletResponse;
+
+
+import java.io.PrintWriter;
+
+@WebServlet("/doggo-servlet")
+public final class DoggoRedditServlet extends HttpServlet {
+
+  final String COMMENT_TEXT = "commentText";
+  final String TIMESTAMP = "timestamp";
+  final String IMAGE_URL = "imageUrl";
+  final String FEED_POST = "FeedPost";
+
+  @Override
+  public void doPost(HttpServletRequest request, HttpServletResponse response) throws IOException {
+
+    Entity postEntity = new Entity(FEED_POST);
+    long timestamp = System.currentTimeMillis();
+    String uploadURL = getUploadedFileUrl(request, "image");
+
+    postEntity.setProperty(COMMENT_TEXT, request.getParameter(COMMENT_TEXT));
+    postEntity.setProperty(TIMESTAMP, timestamp);
+    postEntity.setProperty(IMAGE_URL, uploadURL);
+
+    DatastoreService datastore = DatastoreServiceFactory.getDatastoreService();
+    datastore.put(postEntity);
+
+    response.sendRedirect("/index.html");
+  }
+
+  @Override
+  public void doGet(HttpServletRequest request, HttpServletResponse response) throws IOException {
+    
+    Query query = new Query(FEED_POST).addSort(TIMESTAMP, SortDirection.DESCENDING);
+
+    DatastoreService datastore = DatastoreServiceFactory.getDatastoreService();
+    PreparedQuery results = datastore.prepare(query);
+
+    List<FeedPost> posts = new ArrayList<>();
+    for (Entity entity : results.asIterable()) {
+      
+      String commentText = (String) entity.getProperty(COMMENT_TEXT);
+      String imageUrl = (String) entity.getProperty(IMAGE_URL);
+      long timestamp = (long) entity.getProperty(TIMESTAMP);
+      long id = entity.getKey().getId();
+
+      FeedPost post = new FeedPost(commentText, imageUrl, timestamp);
+      posts.add(post);
+    }
+
+    Gson gson = new Gson();
+
+    response.setContentType("application/json;");
+    response.getWriter().println(gson.toJson(posts));
+  }
+
+  private String getUploadedFileUrl(HttpServletRequest request, String formInputElementName) {
+    BlobstoreService blobstoreService = BlobstoreServiceFactory.getBlobstoreService();
+    Map<String, List<BlobKey>> blobs = blobstoreService.getUploads(request);
+    List<BlobKey> blobKeys = blobs.get(formInputElementName);
+
+    // User submitted form without selecting a file, so we can't get a URL. (dev server)
+    if (blobKeys == null || blobKeys.isEmpty()) {
+      return null;
+    }
+
+    // Our form only contains a single file input, so get the first index.
+    BlobKey blobKey = blobKeys.get(0);
+
+    // User submitted form without selecting a file, so we can't get a URL. (live server)
+    BlobInfo blobInfo = new BlobInfoFactory().loadBlobInfo(blobKey);
+    if (blobInfo.getSize() == 0) {
+      blobstoreService.delete(blobKey);
```

Comment:
can we prevent these from being stored in the first place? How does this case end up happening?

Reply:
Do you mean we should handle it client side where case 1: form no image, send to pure text comment servlet and case 2: form with image sent to /blobstore-image-upload first?

Reply:
Removed this check by making the image required.